### DISCUSSION TITLE
F-Droid submission

### DISCUSSION THREAD

**Discussion by offa** [gokadzev/Musify#165] [Ideas]
### Describe your suggested feature

It would be great to get this app through the official F-Droid repository.

The fastlane metadata is available already :+1:.

### Additional information

- [Inclusion Policy](https://f-droid.org/en/docs/Inclusion_Policy/)
- [How to submit](https://f-droid.org/en/contribute/#submit-applications)

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

**Comment by offa** [comment]
ping @gokadzev

**Reply by gokadzev** [reply]
> ping @gokadzev 

@offa Hello,  do i need to do anything else than removing in-built app updater?

**Reply by offa** [reply]
If you agree to inclusion on F-droid and it's policy I'm going to file a packaging request. According to the policy in-app updates aren't possible, but I guess you can provide a build flavour for F-Droid without it and keep it in the default. I'm not an expert in this topic though :-)

**Reply by gokadzev** [reply]
> If you agree to inclusion on F-droid and it's policy I'm going to file a packaging request. According to the policy in-app updates aren't possible, but I guess you can provide a build flavour for F-Droid without it and keep it in the default. I'm not an expert in this topic though :-)

do i need to change things before submission or it will show us what we need to remove?

**Reply by offa** [reply]
I'm not an expert in this matter, but I can file a request for packaging and we'll see.

**Reply by gokadzev** [reply]
> I'm not an expert in this matter, but I can file a request for packaging and we'll see.

okay, let's try it

**Reply by offa** [reply]
rfp: https://gitlab.com/fdroid/rfp/-/issues/2379

**Reply by gokadzev** [reply]
@offa is there any problem there? https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765

**Reply by offa** [reply]
Not a real problem, I'm just not familiar with flutter builds :-)

It seems to install dependencies correctly but finally fails with

```
ERROR: Could not build app com.gokadzev.musify: No apks match build/com.gokadzev.musify/build/app/outputs/flutter-apk/app-release.apk
```

[This step](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765/diffs#c062ab59d1f22388b1f054c45096fe8fee46aa88_0_30) seems to fail. Do you have an idea?

**Reply by gokadzev** [reply]
> Not a real problem, I'm just not familiar with flutter builds :-)
> 
> It seems to install dependencies correctly but finally fails with
> 
> ```
> ERROR: Could not build app com.gokadzev.musify: No apks match build/com.gokadzev.musify/build/app/outputs/flutter-apk/app-release.apk
> ```
> 
> [This step](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765/diffs#c062ab59d1f22388b1f054c45096fe8fee46aa88_0_30) seems to fail. Do you have an idea?

if it rebuilds the app from source as a release, the problem can be release key/keystore/signing, which of course isn't public. maybe it gets keystore from somewhere else? i have no idea

**Reply by gokadzev** [reply]
also versionCode is wrong in the submission, it should be 48 instead of 483

**Comment by offa** [comment]
> if it rebuilds the app from source as a release, the problem can be release key/keystore/signing, which of course isn't public. maybe it gets keystore from somewhere else? i have no idea

Yes, builds are from source. F-Droid build should do signing automatically (… as far as I know).

> also versionCode is wrong in the submission, it should be 48 instead of 483

Fixed, thanks.

**Comment by offa** [comment]
Build is working now. There's one think left: Is there a way to remove or disable the update (eg. a fdroid build flavour?)

**Reply by gokadzev** [reply]
I'm trying to find any information about it but currently i have nothing. i hope i will be able to remove it only for fdroid builds

**Comment by gokadzev** [comment]
@offa solved:

updater file (which is empty/disabled by default): https://github.com/gokadzev/Musify/blob/master/lib/services/update_manager.dart
check updates button (which is hidden by default and even if it was visible, it would do nothing): https://github.com/gokadzev/Musify/blob/master/lib/screens/more_page.dart#L479

**Reply by offa** [reply]
I'm not a flutter dev, but browsing through other flutter based recipes I found some ways which could help you:

- Using some kind of define: [recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/at.finderlein.noe.yml#L133) and [implementation](https://github.com/finderlein/noefinderlein_flutter/blob/5d0c41f5098162475378a4debf387185b4529aba/android/app/build.gradle#L87)
- Using a build flavour: [recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/cc.narumi.chaldea.fdroid.yml#L414) and [implementation](https://github.com/chaldea-center/chaldea/commit/86e9afb3f9f27eb0930ec1121fbe49b69ef427b1)

(Both projects are just examples)

So you can keep the updater but disable it for fdroid.

**Reply by gokadzev** [reply]
> I'm not a flutter dev, but browsing through other flutter based recipes I found some ways which could help you:
> 
> * Using some kind of define: [recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/at.finderlein.noe.yml#L133) and [implementation](https://github.com/finderlein/noefinderlein_flutter/blob/5d0c41f5098162475378a4debf387185b4529aba/android/app/build.gradle#L87)
> * Using a build flavour: [recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/cc.narumi.chaldea.fdroid.yml#L414) and [implementation](https://github.com/chaldea-center/chaldea/commit/86e9afb3f9f27eb0930ec1121fbe49b69ef427b1)
> 
> (Both projects are just examples)
> 
> So you can keep the updater but disable it for fdroid.

the first one is for native modules, i know the second one but i didn't want to use package_info_plus.

with my way updater is still available but not by default (you need to run bash file and then build), when fdroid will build it from source it's not available but when i run bash file and then build, it becomes available. (maybe i didn't explain very well but to say briefly updater is removed for fdroid builds only)


**Edit: i was wrong, it's possible without package_info_plus. ref: https://github.com/ImranR98/Obtainium/pull/236/files**

**Comment by gokadzev** [comment]
any progress about this?

**Reply by offa** [reply]
The build is green and [PR](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765) waits for review.

**Comment by gokadzev** [comment]
@offa Hello, is it okay for you if i change current "no updates" implementation to the similar one(https://github.com/ImranR98/Obtainium/pull/236/files)? it looks more convenient and logical implementation for me. mainly i'm asking you because for fdroid it gives us different command than current one is (flutter build --flavor fdroid -t lib/main_fdroid.dart), so i guess pr will need changes

**Reply by offa** [reply]
No problem, let me know once I can update the PR. :+1:

**Reply by gokadzev** [reply]
> No problem, let me know once I can update the PR. +1

Done, it works.

**Reply by offa** [reply]
Build updated, but `lib/main_fdroid.dart` isn't available in the 5.0.0 release. Could you create a new release with the file included so I can pick it up in the build?

Is this line correct btw? 

```
$$flutter$$/bin/flutter build apk --flavor fdroid -t lib/main_fdroid.dart
```

I have dropped the `--release` flag, as listed above.

**Reply by gokadzev** [reply]
> Build updated, but `lib/main_fdroid.dart` isn't available in the 5.0.0 release. Could you create a new release with the file included so I can pick it up in the build?
> 
> Is this line correct btw?
> 
> ```
> $$flutter$$/bin/flutter build apk --flavor fdroid -t lib/main_fdroid.dart
> ```
> 
> I have dropped the `--release` flag, as listed above.

I'm not sure about dropping --release flag. About the tag problem i will release 5.1.0

**Reply by offa** [reply]
Ok, then I'm going to add it again and we'll see :-)

**Comment by gokadzev** [comment]
Strange... several hours opened PRs of new apps are accepted/merged but our is delayed

**Reply by offa** [reply]
There's quite a backlog for reviewers, I hope they can have a look soon.

**Reply by offa** [reply]
Update: I'm working on the reproducible build at the moment, with some signature issues though ([context](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765#note_1359440460)).

**Reply by offa** [reply]
@gokadzev is it possible to get a signed fdroid flavour build for the current release so I can add the signature for a reproducible build?

**Reply by gokadzev** [reply]
> @gokadzev is it possible to get a signed fdroid flavour build for the current release so I can add the signature for a reproducible build?

Done, check: https://github.com/gokadzev/Musify/releases/tag/5.1.0

**Comment by gokadzev** [comment]
@offa i checked our PR, as i understand we need to remove specific flutter version and use the latest version right? if that's so i will remove it from pubspec and you will change srclibs: - flutter@3.7.11 to srclibs: - flutter@stable

**Reply by offa** [reply]
You mean [this discussion](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12765#note_1374458532)? I guess the hard coded flutter version is the problem as it needs manual updated once it changes. I've chosen a somewhat up to date version to get it still needs manual update from time to time.

So either adding flutter as a Git submodule ([example](https://github.com/agersant/polaris-android)) or getting the version from the project automatically would be good. But pubspec is a good idea, it contains the info already – maybe we can use it somehow.

**Reply by offa** [reply]
#### Update: 

Solved, there's a good solution using the pubspec file – without changes on your side! :+1:

**Reply by gokadzev** [reply]
> #### Update:
> Solved, there's a good solution using the pubspec file – without changes on your side! +1

Perfect 🎉

**Reply by offa** [reply]
I have opened a PR to prevent different flutter versions between GH and F-Droid (#231). This isn't necessary for the current release or RFP, but for future versions.

**Reply by gokadzev** [reply]
> I have opened a PR to prevent different flutter versions between GH and F-Droid (#231). This isn't necessary for the current release or RFP, but for future versions.

Thanks, merged.

**Comment by gokadzev** [comment]
@offa New version with changes is released (as well as fdroid version).

**Reply by offa** [reply]
Thanks, I have updated the F-Droid PR. Hopefully it's merged soon.

**Reply by offa** [reply]
Update: PR has been merged and should appear on F-Droid within the next days – including a reproducible build :tada:.

**Reply by gokadzev** [reply]
> Update: PR has been merged and should appear on F-Droid within the next days – including a reproducible build tada.

Thanks a lot! 🎉 I appreciate your help

**Comment by licaon-kter** [comment]
No `-fdroid.apk` for https://github.com/gokadzev/Musify/releases/tag/5.3.0 ?

ref: https://gitlab.com/fdroid/fdroiddata/-/jobs/4331465982#L1116

**Reply by gokadzev** [reply]
Sorry, I forgot. It's added now.

**Reply by gokadzev** [reply]
@offa Hello, do we need to do anything after build failed once? because as i see new version isn't added yet

**Reply by licaon-kter** [reply]
No, I've left it alone so it will be retried ;)

**Reply by licaon-kter** [reply]
this faq applies to updates too: https://gitlab.com/fdroid/wiki/-/wikis/FAQ#how-long-does-it-take-for-my-app-to-show-up-on-website-and-client :D
